specification/transport/udp/udp.tex

```diff
@@ -0,0 +1,337 @@
+
+bool end_of_transfer
+# True if this is the last frame of a multi-frame transfer, or a single-frame transfer.
+
+uint16 user_data
+# Opaque application-specific data with user-defined semantics.
+# Generic implementations should emit zero and ignore this field upon reception.
+
+uint8[2] header_crc16_big_endian
+# CRC-16/CCITT-FALSE of the preceding serialized header data in the big endian byte order.
+# Application of the CRC function to the entire header shall yield zero, otherwise the header is malformed.
+
+@assert _offset_ / 8 == {24}
+@sealed     # The payload data follows.
+\end{minted}
+\end{samepage}
+
+The header CRC function is \textbf{CRC-16/CCITT-FALSE};
+refer to section~\ref{sec:appendix_crc16ccitt_false} for further information.
+
+\begin{remark}
+    Certain states provided in the header duplicate information that is already available in the IP header
+    or the multicast group address.
+    This is done for reasons of unification of the header format with other standard transport layer definitions,
+    and to simplify the access to the transfer parameters that otherwise would be hard to reach above the
+    network layer, such as the DSCP value.
+    The latter consideration is particularly important for forwarding nodes.
+\end{remark}
+
+\subsection{Transfer payload}
+
+After the transfer payload is constructed but before it is scheduled for transmission over the network,
+it is appended with the transfer CRC field.
+The transfer CRC function is \textbf{CRC-32C} (section~\ref{sec:appendix_crc32c}),
+and its value is serialized in the little-endian byte order.
+The transfer CRC function is applied to the entire transfer payload and only transfer payload.
+
+The transfer CRC is provided for all transfers,
+including single-frame transfers and transfers with an empty payload\footnote{%
+    This provides end-to-end integrity protection for the transfer payload.
+}.
+An implementation receiving a transfer should verify the correctness of its transfer CRC.
+
+\begin{remark}
+    From the perspective of the multi-frame segmentation logic, the transfer CRC field is part of the transfer payload.
+    From the definition of the header format it follows that the transfer CRC can only be found at the end of
+    the packet if the \verb|end_of_transfer| bit is set,
+    unless the transfer CRC field has spilled over to the next frame
+    (in which case the frame would contain only the transfer CRC itself or the tail thereof).
+\end{remark}
+
+\subsection{Maximum transmission unit}
+
+In this section, the maximum transmission unit (MTU) is defined as the maximum size of a UDP/IP datagram payload.
+
+This specification does not restrict the MTU of the underlying transport.
+It is recommended, however, to avoid MTU values less than 508~bytes to allow reduced (simplified) implementations
+of the protocol that do not require large transfer payloads to omit support for multi-frame transfers.
```

Comment:
What about IP fragmentation and reassembly? Do we require and/or recommend that IP framing be used in lieu of Cyphal framing? Do we prohibit it? It seems complicated, if hand-writing a Cyphal/UDP/IP stack, to have to support both.

Reply:
I am inclined to prohibit multipart IP transfers as they are indeed redundant in the presence of Cyphal framing, but wouldn't that be overreach on our part? We define the protocol in terms of UDP/IP, and, strictly speaking, most of the things that happen underneath are not our responsibility.

Reply:
This continues to vex me. It's important, for the smallest and/or most critical nodes, that the amount of software required to run Cyphal is minimised. This means the network designer has to chose between ip framing or Cyphal. I'd like to take some stance on this and I'm leaning towards prohibiting multi-part Cyphal messages on IP networks. What reason do we have to go the other way and prohibit ethernet framing in favor of Cyphal?